File: cpp/match.cpp
#include "structs.hpp"

 

double DISTANCE_MATCH = 0.75;
double LENGTH_MATCH = 0.05;
double ANGLE_MATCH = 0.05;
double MIN_MATCH = 5;
string ext = ".bin";

bool fingerprint_match(string img_path, string clear_path, string sample_path){

	cv::Mat image = (256 - cv::imread(clear_path)) - cv::imread(img_path);

	cv::Mat ROI = cv::Mat::ones(cv::Size(image.size[0], image.size[1]), 0);

	double maximum;
	double minimum;

	minMaxLoc(image, &minimum, &maximum, NULL, NULL);
	image = (255/(maximum-minimum)) * (image-minimum);

	vector<cv::KeyPoint> keypoints_image;
	cv::Mat descriptors_image; 
	cv::SIFT::create()->detectAndCompute(image, cv::noArray(), keypoints_image, descriptors_image);

	if(keypoints_image.size() < MIN_MATCH)
		return false;

	cv::Ptr<cv::BFMatcher> BFM = cv::BFMatcher::create();

	for(auto file : recursive_directory_iterator(sample_path)){
		if(((string)file.path()).substr(((string)file.path()).size()-4, 4) == ext){

			vector<cv::KeyPoint> keypoints_match;
			cv::Mat descriptors_match; 
			
			cv::FileStorage store(file.path(), cv::FileStorage::READ);

			cv::FileNode n1 = store["keypoints"];
			cv::FileNode n2 = store["descriptors"];
			
			cv::read(n1,keypoints_match);
			cv::read(n2,descriptors_match);
			
			store.release();
			
			if(keypoints_match.size() < MIN_MATCH){
				continue;
			}
			
			vector<vector<cv::DMatch>> points;
			BFM->knnMatch(descriptors_image, descriptors_match, points, 2);
			
			vector<match> matches;
			for (int i = 0; i < points.size(); i++) {
				cv::DMatch match_1 = points[i][0];
				if (match_1.distance < DISTANCE_MATCH * points[i][1].distance) {
					matches.push_back({ keypoints_image[match_1.queryIdx].pt,
										keypoints_match[match_1.trainIdx].pt});
				}
			}
			cout << matches.size() << endl;
			if(matches.size() < MIN_MATCH){
				continue;
			}

			vector<angle> angles;

			set<match> set(matches.begin(), matches.end());
			matches.assign(set.begin(), set.end());

			for(int j = 0; j < matches.size(); j++){
				match match_1 = matches[j];
				for(int k = j+1; k < matches.size(); k++){
					match match_2 = matches[k];

					double vec_1 [2] = {match_1.p1.x - match_2.p1.x,
										match_1.p1.y - match_2.p1.y};
					double vec_2 [2] = {match_1.p2.x - match_2.p2.x,
										match_1.p2.y - match_2.p2.y};

					double length_1 = sqrt(pow(vec_1[0],2) + pow(vec_1[1],2));
					double length_2 = sqrt(pow(vec_2[0],2) + pow(vec_2[1],2));

					if (1 - min(length_1, length_2) / 
							max(length_1, length_2) <= LENGTH_MATCH){
			
						double product = length_1 * length_2;
						double vec [2] = {};
						angles.push_back(angle(
							M_PI / 2 + asin((vec_1[0] * vec_2[0] + vec_1[1] * vec_2[1]) / product),
							acos((vec_1[0] * vec_2[1] - vec_1[1] * vec_2[0]) / product),
							match_1,
							match_2
							));
					}
				}
			}

			if(angles.size()<MIN_MATCH)
				continue;

			int count = 0;
			for(int j = 0; j < angles.size(); j++){
				angle angle_1 = angles[j];
				for(int k = j+1; k < angles.size(); k++){
					angle angle_2 = angles[k];

					if (1 - min(angle_1.sin, angle_2.sin) / 
							max(angle_1.sin, angle_2.sin) <= ANGLE_MATCH &&
						1 - min(angle_1.cos, angle_2.cos) /
							max(angle_1.cos, angle_2.cos) <= ANGLE_MATCH){

							count += 1;

						if (count >= MIN_MATCH)
						return true;
					}
				}
			}
			cout << count << endl;
		}
	}
	return false;
}

int main(int argc, char** argv){
	cout << fingerprint_match(argv[1], argv[2], argv[3]);
	return 0;
}

File: ref/match.py
from time import time

import cv2 as cv
import numpy as np

DISTANCE_MATCH = 0.75
LENGTH_MATCH = 0.05
ANGLE_MATCH = 0.05
MIN_MATCH = 5


def match(image_path: str, clear_path: str, sample_path) -> bool:
    image: np.ndarray = cv.imread(image_path, 0) - cv.imread(clear_path, 0)

    minimum: int = np.min(image)
    maximum: int = np.max(image)
    tmp = 255 / (maximum - minimum)
    image = np.uint8(np.round(tmp * image - minimum * tmp))

    keypoints_1: tuple
    descriptors_1: np.ndarray
    keypoints_1, descriptors_1 = cv.SIFT_create().detectAndCompute(image, None)

    if len(keypoints_1) < MIN_MATCH:
        return False

    matcher = cv.BFMatcher()
    with np.load(sample_path, allow_pickle=True) as samples:
        for name in samples:
            keypoints_2: np.ndarray
            descriptors_2: np.ndarray
            keypoints_2, descriptors_2 = samples[name]

            if len(keypoints_2) < MIN_MATCH:
                continue

            descriptors_2 = np.stack(descriptors_2)
            points: tuple = matcher.knnMatch(descriptors_1, descriptors_2, k=2)

            if len(points) < MIN_MATCH:
                continue

            matchs = []
            for match_1, match_2 in points:
                if match_1.distance < DISTANCE_MATCH * match_2.distance:
                    matchs.append((keypoints_1[match_1.queryIdx].pt,
                                   keypoints_2[match_1.trainIdx]))

            matchs = list(set(matchs))

            if len(matchs) < MIN_MATCH:
                continue

            angles = []
            for i, match_1 in enumerate(matchs):
                for match_2 in matchs[i + 1:]:
                    vec_1 = (match_1[0][0] - match_2[0][0],
                             match_1[0][1] - match_2[0][1])
                    vec_2 = (match_1[1][0] - match_2[1][0],
                             match_1[1][1] - match_2[1][1])

                    length_1 = np.sqrt(vec_1[0]**2 + vec_1[1]**2)
                    length_2 = np.sqrt(vec_2[0]**2 + vec_2[1]**2)

                    if 1 - min(length_1, length_2) / max(
                            length_1, length_2) <= LENGTH_MATCH:
                        product = length_1 * length_2
                        angles.append(
                            (np.pi / 2 + np.arcsin(
                                (vec_1[0] * vec_2[0] + vec_1[1] * vec_2[1]) /
                                product),
                             np.arccos(
                                 (vec_1[0] * vec_2[1] - vec_1[1] * vec_2[0]) /
                                 product)))

            if len(angles) < MIN_MATCH:
                continue

            for i, angle_1 in enumerate(angles):
                count = 0
                for j, angle_2 in enumerate(angles):
                    if i == j:
                        continue

                    if 1 - min(angle_1[0], angle_2[0]) / max(
                            angle_1[0], angle_2[0]) <= ANGLE_MATCH and 1 - min(
                                angle_1[1], angle_2[1]) / max(
                                    angle_1[1], angle_2[1]) <= ANGLE_MATCH:
                        count += 1

                    if count >= MIN_MATCH:
                        return True

        return False


if __name__ == '__main__':
    start = time()
    print(
        match('fingerprints/fingerprint.png', 'fingerprints/clear.png',
              'fingerprints/finger-2.npz'))
    print(f"Time: {time() - start}")
